Anatomy of a cmdlet definition (#77)
PowerShell/Community-Blog: Posts/2021/10/crescendo-configuration.md

Summary: In this post I take a closer look at a cmdlet definition in the Crescendo configuration file.
featured_image: https://devblogs.microsoft.com/powershell-community/wp-content/uploads/sites/69/2021/09/Crescendo.png
tags: Crescendo, cmdlet, json, configuration
username: sewhee@microsoft.com
---
In my [previous post][3], I looked at the details of a Crescendo output handler from my
[VssAdmin][7] module. In this post, I explain the details of a cmdlet definition in the Crescendo
JSON configuration file.

## The purpose of the configuration

The structure for the interface of a cmdlet is a reasonably predictable thing.

- The cmdlet uses a standard _Verb-Noun_ format
- The cmdlets take input using sets of parameters
- Cmdlets that make changes to the system support `-Confirm` and `-WhatIf` parameters

The pattern of the script code to support these fits a template.

The more difficult part of the cmdlet is in the code that does the work. Crescendo separates the
functional code (the output handler) from the cmdlet interface code. The Crescendo configuration
file defines the interfaces of cmdlets that you want Crescendo to create.

> The Crescendo configuration file is a JSON file containing a array of cmdlet definitions. JSON

Reviewer: ```suggestion
The Crescendo configuration file is a JSON file containing an array of cmdlet definitions. JSON
```

> provides a expressive, structured syntax for defining the properties of object. But so does,

Reviewer: ```suggestion
provides an expressive, structured syntax for defining the properties of objects. But so does,
```

PowerShell. So why use JSON and not a PowerShell data (PSD1) file? The answer is simple: schema!
Unlike PowerShell's PSD1 files, JSON supports a schema. Having a schema ensures that the syntax of
your definition is correct. And with tools like Visual Studio Code (VS Code), the schema provides
IntelliSense, making it easier to author.

## Defining a cmdlet interface

> The structure of a cmdlet definition can be divided into three categories of properties in the JSON

Reviewer: ```suggestion
The structure of a cmdlet definition can be divided into three property categories in the JSON
```

file:

- Required properties
  - **Verb**
  - **Noun**
  - **OriginalName**
  - **OriginalCommandElements[]**
  - **OutputHandlers[]**
- As-required properties
  - **DefaultParameterSetName**
  - **Parameters[]**
- Nice-to-have properties
  - "Help" properties like **Description**, **Usage**, and **Examples[]**

You might notice that defining **Parameters** is optional. This is not uncommon. In my VssAdmin
module, the cmdlets `Get-VssProvider`, `Get-VssVolume`, and `Get-VssWriter` do not have parameters.

> These simple cmdlet don't require any input to return the requested information.

Reviewer: ```suggestion
These simple cmdlets don't require any input to return the requested information.
```

Let's take a closer look at a simple cmdlet definition.

- The **Verb** and **Noun** properties form the name of the cmdlet.
- The **OriginalName** property contains the path to the native command that the cmdlet runs to get
  the output.

> - The **OriginalCommandElements** is an array of strings that passed to the native command as

Reviewer: ```suggestion
- The **OriginalCommandElements** is an array of strings that are passed to the native command as
```
